File: tests/Web.AcceptanceTests/Pages/WorkoutDetailPage.cs
using System.Text.RegularExpressions;

namespace LiftAndShift.Web.AcceptanceTests.Pages;

public class WorkoutDetailPage(IPage page) : BasePage(page)
{
    public override string PagePath => $"{BaseUrl}/workout";

    public Task GotoWorkout(int id) => Page.GotoAsync($"{BaseUrl}/workout/{id}");

    public Task AssertOnWorkoutDetailPage() => Assertions.Expect(Page).ToHaveURLAsync(new Regex(@"/workout/\d+$"));

    public Task AssertExerciseVisible(string name) =>
        Assertions.Expect(Page.Locator("article header strong", new() { HasText = name })).ToBeVisibleAsync();

    public Task AssertStatus(string status) => Assertions.Expect(Page.Locator("hgroup p")).ToContainTextAsync(status);

    public Task ClickMarkComplete() => Page.Locator("button:has-text('Mark Complete')").ClickAsync();

    public Task ClickDuplicate() => Page.Locator("button:has-text('Duplicate as New Draft')").ClickAsync();

    public Task AssertMarkCompleteButtonHidden() =>
        Assertions.Expect(Page.Locator("button:has-text('Mark Complete')")).ToHaveCountAsync(0);
}

File: src/Web/ClientApp/e2e/pages/workout-detail.page.ts
import { expect } from '@playwright/test';
import { BasePage, baseUrl } from './base-page';

export class WorkoutDetailPage extends BasePage {
  get pagePath(): string {
    return `${baseUrl()}/workout`;
  }

  gotoWorkout(id: number) {
    return this.page.goto(`${baseUrl()}/workout/${id}`);
  }

  assertOnWorkoutDetailPage() {
    return expect(this.page).toHaveURL(/\/workout\/\d+$/);
  }

  assertExerciseVisible(name: string) {
    return expect(this.page.locator('article header strong', { hasText: name })).toBeVisible();
  }

  assertStatus(status: string) {
    return expect(this.page.locator('hgroup p')).toContainText(status);
  }

  clickMarkComplete() {
    return this.page.locator("button:has-text('Mark Complete')").click();
  }

  clickDuplicate() {
    return this.page.locator("button:has-text('Duplicate as New Draft')").click();
  }

  assertMarkCompleteButtonHidden() {
    return expect(this.page.locator("button:has-text('Mark Complete')")).toHaveCount(0);
  }
}
